### DISCUSSION TITLE
Memories

### DISCUSSION THREAD

**Discussion by ceicke** [AllYourBot/hostedgpt#512] [General]
I saw that my HostedGPT instance created one memory. It's a nice memory that it created, but I was wondering how this works? I saw that ChatGPT has the same thing. When are memories created? How are they used in the prompts?

**Comment by keithschacht** [comment]
To be honest, I haven’t found it to be super useful in practice. But basically the LLM is given an optional function to call if the user just told it something which is the kind of thing they would expect their assistant to remember about them. Then all memories are appended as part of the system prompt here: https://github.com/AllYourBot/hostedgpt/blob/main/app/services/ai_backend.rb#L94

It’s a naive implementation that wouldn’t scale if it started to remember too many things. It also doesn’t account for naturally forgetting information.

I’ve found it a little useful in voice mode. I told it my wife’s name, my wife’s email address, so now I can simply make reference to her by name and it knows who I’m talking about. So it works well to teach it basic facts about yourself.

**Comment by mattlindsey** [comment]
I just found Assistant Memories pretty useful for remembering a fact when doing some research: the date a building was built and who owns it now.    That memory allowed the AI to make other connections and seemed to 'unlock' some knowledge about the connection that it couldn't find originally.

I might add an Issue to allow deleting individual memories with a little trash can beside each one, like ChatGPT has.

**Reply by keithschacht** [reply]
I like the idea! I saw the PR and will review it soon.

The one other thing that I’ve noticed ChatGPT does is a simple output within your interactive chat session that says “memory updated”.

That would be nice for us to support too, at some point. If I recall correctly, I think we already have multiple `div` for every tool call within the front-end UI, we just add hidden classes to them when we realize they are tool calls? That’s my rough memory. So there may be a way to have a user-visible output from tool calls so that the Memory call could notify the user “Memory updated”

Oh, oh, actually, an easier way we could implement it — but less elegant — is to simply modify the system message for the Memory tool to include the instruction “any time you save something to memory be sure to let the user know you did so.” Then I think it would just “be smart” and incorporate this into it’s next reply.
